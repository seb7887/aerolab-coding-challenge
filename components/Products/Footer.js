import React from 'react';
import styled from 'styled-components';

import Quantity from '../shared/Quantity';
import Pagination from '../shared/Pagination';

import config from '../../config';

const Container = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  padding-bottom: 0.8rem;
  border-bottom: 1px solid ${props => props.theme.grey};

  @media (max-width: 770px) {
    justify-content: flex-end;
  }
`;

const Footer = ({ page, total }) => {
  const { perPage } = config;
  const quantity = total > perPage ? (parseInt(page, 10) + 1) * perPage : total;

  return (
    <Container>
      <Quantity quantity={quantity} total={total} />
      <Pagination page={page} total={total} perPage={perPage} />
    </Container>
  );
};

export default Footer;
