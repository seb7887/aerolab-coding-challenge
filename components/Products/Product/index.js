import React from 'react';

import BuyIcon from './BuyIcon';
import Coin from '../../shared/Coin';

import {
  Container,
  Img,
  Content,
  OverlayBlock,
  RedeemButton,
  NeedCoins
} from './style';

const Product = ({ product, userPoints, redeemProduct }) => (
  <Container data-testid='product'>
    <Content>
      <div className='buy'>
        {userPoints < product.cost ? (
          <NeedCoins data-testid='need-coins'>
            You need {product.cost - userPoints} <Coin />
          </NeedCoins>
        ) : (
          <BuyIcon color='orange' />
        )}
      </div>
      <Img src={product.img.url} alt={product.name} />
      <div className='description'>
        <span className='category' data-testid='category'>
          {product.category}
        </span>
        <span className='name' data-testid='name'>
          {product.name}
        </span>
      </div>
      {userPoints >= product.cost && (
        <div className='overlay'>
          <div className='buy'>
            <BuyIcon color='white' />
          </div>
          <OverlayBlock>
            <div className='cost-block'>
              <span className='cost' data-testid='cost'>
                {product.cost}
              </span>
              <Coin />
            </div>
            <RedeemButton
              onClick={() => redeemProduct(product._id)}
              data-testid='redeem-button'
            >
              Redeem Now
            </RedeemButton>
          </OverlayBlock>
        </div>
      )}
    </Content>
  </Container>
);

export default Product;
