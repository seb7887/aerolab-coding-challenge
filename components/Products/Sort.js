import React from 'react';

const Sort = ({ total }) => (
  <div>
    <div>
      <span>{total}</span> PRODUCTS
    </div>
  </div>
);

export default Sort;
