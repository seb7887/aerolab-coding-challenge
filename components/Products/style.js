import styled from 'styled-components';

export const ProductsLayout = styled.div`
  padding: 1.5rem;
  display: grid;
  grid-template-areas:
    'top top'
    'filters grid'
    'bottom bottom';

  #top {
    grid-area: top;
  }

  #filters {
    grid-area: filters;
  }

  #grid {
    grid-area: grid;
  }

  #bottom {
    grid-area: bottom;
  }

  @media (max-width: 770px) {
    padding: 1rem;
  }

  @media (max-width: 650px) {
    padding: 0.5rem;
  }
`;
