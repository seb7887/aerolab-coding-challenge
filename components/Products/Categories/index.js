import React from 'react';
import Link from 'next/link';
import shortid from 'shortid';

import { Container } from './style';

const categories = [
  'all',
  'gaming',
  'laptops',
  'tablets',
  'cameras',
  'audio',
  'tv',
  'drones',
  'accessories',
  'home'
];

const Categories = ({ category, sort }) => (
  <Container>
    <h3>Categories</h3>
    <>
      {categories.map(name => {
        const isCurrent = name === category;
        return (
          <Link
            prefetch
            href={{
              pathname: '/',
              query: {
                category: name,
                sort,
                page: 0
              }
            }}
            key={shortid.generate()}
          >
            <a className={isCurrent ? 'current' : ''}>{name}</a>
          </Link>
        );
      })}
    </>
  </Container>
);

export default Categories;
