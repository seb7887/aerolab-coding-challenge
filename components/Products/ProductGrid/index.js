import React from 'react';

import Product from '../Product';

import { Grid } from './style';

const ProductGrid = ({ userPoints, products, redeemProduct }) => (
  <Grid>
    {products.map(product => {
      return (
        <Product
          key={product._id}
          product={product}
          userPoints={userPoints}
          redeemProduct={redeemProduct}
        />
      );
    })}
  </Grid>
);

export default ProductGrid;
