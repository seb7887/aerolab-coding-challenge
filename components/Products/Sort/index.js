import React from 'react';

import Quantity from '../../shared/Quantity';
import Filters from './Filters';
import Pagination from '../../shared/Pagination';

import { Container } from './style';

import config from '../../../config';

const Sort = ({ page, sort, category, total }) => {
  const { perPage } = config;
  const quantity = total > perPage ? (parseInt(page, 10) + 1) * perPage : total;

  return (
    <Container>
      <Quantity quantity={quantity} total={total} />
      <div className='sort'>
        <Filters category={category} sort={sort} />
        <Pagination
          category={category}
          sort={sort}
          page={page}
          total={total}
          perPage={perPage}
        />
      </div>
    </Container>
  );
};

export default Sort;
