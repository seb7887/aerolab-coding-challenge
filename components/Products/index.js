import React from 'react';

import Sort from './Sort';
import Categories from './Categories';
// TODO: saca esto y reemplaza por ProductGrid
import ProductGrid from './ProductGrid';
import Footer from './Footer';

import { ProductsLayout } from './style';

const Products = ({ page, sort, category, products }) => (
  <ProductsLayout>
    <div id='top'>
      <Sort
        category={category}
        sort={sort}
        page={page}
        total={products.total}
      />
    </div>
    <div id='filters'>
      <Categories category={category} sort={sort} />
    </div>
    <div id='grid'>
      <ProductGrid products={products.products} />
    </div>
    <div id='bottom'>
      <Footer
        category={category}
        sort={sort}
        page={page}
        total={products.total}
      />
    </div>
  </ProductsLayout>
);

export default Products;
