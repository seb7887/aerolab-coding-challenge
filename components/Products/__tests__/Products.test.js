import { render } from 'react-testing-library';

import Products from '../index';

jest.mock('next/link', () => {
  return ({ children }) => {
    return children;
  };
});

import { fakeUser, fakeProducts } from '../../../utils/testUtils';

const mockRedeemProduct = jest.fn();
const page = 0;
const sort = 'recent';
const category = 'all';
const products = {
  products: fakeProducts,
  total: fakeProducts.length
};

describe('Products', () => {
  it('renders and matches snapshot', () => {
    const { asFragment } = render(
      <Products
        me={fakeUser}
        page={page}
        sort={sort}
        category={category}
        products={products}
        redeemProduct={mockRedeemProduct}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
