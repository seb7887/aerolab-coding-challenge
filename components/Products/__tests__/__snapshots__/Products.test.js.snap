// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="style__ProductsLayout-sc-1vq3ucx-0 LYoHN"
  >
    <div
      id="top"
    >
      <div
        class="style__Container-sc-1cigh81-0 MGrfv"
      >
        <div
          class="Quantity__Container-sc-1ye5v54-0 kMaZJe"
        >
          <span>
            1 of 1
          </span>
           products
        </div>
        <div
          class="sort"
        >
          <div
            class="Filters__Container-sc-1w8ug50-0 eIrHFy"
          >
            <div
              class="title"
            >
              Sort by:
            </div>
            <div
              class="buttons"
            >
              <a
                class="current"
                data-testid="sort-filter"
              >
                Most recent
              </a>
              <a
                class=""
                data-testid="sort-filter"
              >
                Lower price
              </a>
              <a
                class=""
                data-testid="sort-filter"
              >
                Higher price
              </a>
            </div>
          </div>
          <div
            class="Pagination__Container-j6vdk2-0 lhMuFC"
          />
        </div>
      </div>
    </div>
    <div
      id="filters"
    >
      <div
        class="style__Container-sc-15ofmo2-0 SyiiL"
      >
        <h3>
          Categories
        </h3>
        <a
          class="current"
          data-testid="category"
        >
          all
        </a>
        <a
          class=""
          data-testid="category"
        >
          gaming
        </a>
        <a
          class=""
          data-testid="category"
        >
          laptops
        </a>
        <a
          class=""
          data-testid="category"
        >
          tablets
        </a>
        <a
          class=""
          data-testid="category"
        >
          cameras
        </a>
        <a
          class=""
          data-testid="category"
        >
          audio
        </a>
        <a
          class=""
          data-testid="category"
        >
          tv
        </a>
        <a
          class=""
          data-testid="category"
        >
          drones
        </a>
        <a
          class=""
          data-testid="category"
        >
          accessories
        </a>
        <a
          class=""
          data-testid="category"
        >
          home
        </a>
      </div>
    </div>
    <div
      id="grid"
    >
      <div
        class="style__Grid-sc-15ghthz-0 gxQmBB"
      >
        <div
          class="style__Container-g760h8-0 gPFdvq"
          data-testid="product"
        >
          <div
            class="style__Content-g760h8-5 tttFT"
          >
            <div
              class="buy"
            >
              <img
                alt="Buy"
                src="/static/images/buy-orange.svg"
              />
            </div>
            <img
              alt="iPhone"
              class="style__Img-g760h8-2 frYZGq"
              src="/static/image/image.png"
            />
            <div
              class="description"
            >
              <span
                class="category"
                data-testid="category"
              >
                Phones
              </span>
              <span
                class="name"
                data-testid="name"
              >
                iPhone
              </span>
            </div>
            <div
              class="overlay"
            >
              <div
                class="buy"
              >
                <img
                  alt="Buy"
                  src="/static/images/buy-white.svg"
                />
              </div>
              <div
                class="style__OverlayBlock-g760h8-3 lladji"
              >
                <div
                  class="cost-block"
                >
                  <span
                    class="cost"
                    data-testid="cost"
                  >
                    800
                  </span>
                  <img
                    alt="Coin"
                    class="Coin__Img-sc-1i19xvt-0 cnqtJE"
                    src="/static/images/coin.svg"
                  />
                </div>
                <a
                  class="style__RedeemButton-g760h8-4 cUoLRy"
                  data-testid="redeem-button"
                >
                  Redeem Now
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      id="bottom"
    >
      <div
        class="Footer__Container-sc-1ac1drc-0 eoifZM"
      >
        <div
          class="Quantity__Container-sc-1ye5v54-0 kMaZJe"
        >
          <span>
            1 of 1
          </span>
           products
        </div>
        <div
          class="Pagination__Container-j6vdk2-0 lhMuFC"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
